### PR TITLE
Add experimental.tensor_resize operator

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Introduces a layout-agnostic frontend for Resize
- New operator `experimental.tensor_resize`
- All dimensions can be resized (including channels), the implementation is able to detect unchanged dimensions at the beginning and the end, and treat those as non-spatial dimensions.
- Up to 3 spatial dimensions supported

### Affected modules and functionalities:

- New op (related to resize)

### Key points relevant for the review:

Resize parameter logic

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A
test_tensor_resize.test_resize_upsample_scales_nearest
test_tensor_resize.test_resize_downsample_scales_nearest
test_tensor_resize.test_resize_upsample_sizes_nearest
test_tensor_resize.test_resize_upsample_scales_linear
test_tensor_resize.test_resize_downsample_scales_linear
test_tensor_resize.test_resize_alignment
test_tensor_resize.test_resize_upsample_scales_cubic
test_tensor_resize.test_resize_downsample_scales_cubic
test_tensor_resize.test_resize_upsample_sizes_cubic
test_tensor_resize.test_resize_downsample_sizes_cubic

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3148

### CODE REVIEW

File: dali/operators/image/resize/resize_attr_base.h [hunk 1]
```diff
@@ -103,8 +103,20 @@ void CalculateSampleParams(ResizeParams &params, SmallVector<float, 3> requested
...
+      } else if (alignment[d] == -1) {
+        // keep start of the ROI
+        center = params.src_lo[d];
+      } else if (alignment[d] == 1) {
+        // keep end of the ROI
+        center = params.src_hi[d];
+      } else {
+        DALI_FAIL(make_string("Unsupported alignment value ", alignment[d],
+                              ". Supported values are 0, -1, 1"));
+      }
```
Comment (mzient) [inline dali/operators/image/resize/resize_attr_base.h:119]: To consider:
```suggestion
      } else if (alignment[d] < -1) {
        // keep start of the ROI
        center = params.src_lo[d];
      } else if (alignment[d] > 1) {
        // keep end of the ROI
        center = params.src_hi[d];
      }
```
Having a sign (as opposed to exact value) define the behavior may simplify the client code. The usage of sign is not uncommon (see `strcmp` or the "spaceship" operator `<=>`).

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+When ``axes`` is provided, the scale factor values refer to the axes specified.
+Note: Arguments ``sizes`` and ``scales`` are mutually exclusive.)code",
+                                   {}, true)
+    .AddOptionalArg<int>("alignment", R"code(Determines the position of the ROI 
+when using scales (provided or calculated).
+
+Accepted values are -1 (align with top-left corner), 0 (centered), 1 (align with bottom-right corner).
+By default, 0 (centered) is assumed. Contains as many elements as dimensions provided for sizes/scales.
+If only one value is provided, it is apply to all dimensions.)code",
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:45]: If you go with the sign, adjust the comment accordingly.

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+The resulting output size is calculated as
+``out_size = size_rounding(scale_factor * original_size)``.
+See ``size_rounding`` for a list of supported rounding policies.
+
+When ``axes`` is provided, the scale factor values refer to the axes specified.
+Note: Arguments ``sizes`` and ``scales`` are mutually exclusive.)code",
+                                   {}, true)
+    .AddOptionalArg<int>("alignment", R"code(Determines the position of the ROI 
+when using scales (provided or calculated).
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:41]: I'd add some more explanation.
```

The real output size must be integer and may differ from "ideal" output size calculated as input (or ROI) size multiplied by the scale factor. In that case, the output size is rounded (according to `size_rounding` policy) and the input ROI needs to be adjusted to maintain the scale factor. This parameter defines which point of the ROI should remain constant.
```

File: dali/operators/image/resize/resize_attr_base.h (hunk 1 above)
Comment (mzient) [reply dali/operators/image/resize/resize_attr_base.h:119]: A random thought: how about having an alignment as a `float` 0..1? 0.5 would denote the center.
```C++
center = (1 - alignment[d]) * params.src_lo[d] + alignment[d] * params.src_hi[d]
```

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+By default, all dimensions are assumed.)code",
+        std::vector<int>{})
+    .AddOptionalArg<std::string>("size_rounding",
+                                 R"code(Determines the rounding policy when using scales.
+
+Possible values are:
+* | ``"round"`` - Rounds the resulting size to the nearest integer value, with halfway cases rounded away from zero.
+* | ``"truncate"`` - Discards the fractional part of the resulting size.
+* | ``"ceil"`` - Rounds up the resulting size to the next integer value.)code",
+                                 "truncate")
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:60]: why not `"round"`? Don't we round in normal Resize?

Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:60]: true, by default we round. I'll change it

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+This point is calculated as ``center = (1 - alignment) * roi_start + alignment * roi_end``.
+Alignment 0.0 denotes alignment with the start of the ROI, 0.5 with the center of the region, and 1.0 with the end.
+Note that when ROI is not specified, roi_start=0 and roi_end=size is assumed.
+
+When using 0.5 (default), the resize operation has flip invariant properties (flipping after resizing is
+mathematically equivalent to resizing after flipping).
+
+Contains as many elements as dimensions provided for sizes/scales. If only one value is provided, it is
+applied to all dimensions.)code",
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:56]: Now, with multiple paragraphs above, it's hard to infer the subject.
```suggestion
The value of this argument contains as many elements as dimensions provided for sizes/scales.
If only one value is provided, it is applied to all dimensions.)code",
```

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+when using scales (provided or calculated).
+
+The real output size must be integer and may differ from "ideal" output size calculated as input
+(or ROI) size multiplied by the scale factor. In that case, the output size is rounded (according
+to `size_rounding` policy) and the input ROI needs to be adjusted to maintain the scale factor.
+This parameter defines which relative point of the ROI should retain its position in the output.
+
+This point is calculated as ``center = (1 - alignment) * roi_start + alignment * roi_end``.
+Alignment 0.0 denotes alignment with the start of the ROI, 0.5 with the center of the region, and 1.0 with the end.
+Note that when ROI is not specified, roi_start=0 and roi_end=size is assumed.
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:50]:
```suggestion
Note that when ROI is not specified, roi_start=0 and roi_end=input_size is assumed.
```

File: dali/operators/generic/resize/tensor_resize.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+  void PrepareParams(const ArgumentWorkspace &ws, const TensorListShape<> &input_shape) {
+    int nsamples = input_shape.num_samples();
+    resize_attr_.PrepareResizeParams(spec_, ws, input_shape);
+
+    if (NumSpatialDims() < 1) {
+      throw std::invalid_argument(
+          "TensorResize expects at least one spatial dimension to be resized");
+    }
+
```
Comment (mzient) [inline dali/operators/generic/resize/tensor_resize.h:55]: Perhaps for the time being, you can simply report 1D as not implemented and keep the operator in `experimental`.
```suggestion

    if (NumSpatialDims() == 1) {
      throw std::invalid_argument("1D resizing is not implemented.");
    }

```

File: dali/operators/image/resize/tensor_resize_attr.h [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace dali {
+
+class DLL_PUBLIC TensorResizeAttr {
+ public:
+  explicit TensorResizeAttr(const OpSpec &spec) {
+    has_axes_ = spec.HasArgument("axes");
+    if (has_axes_) {
+      axes_ = spec.GetRepeatedArgument<int>("axes");
+    }
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.h:33]: Isn't it repeated at `TensorResizeAttr::SetFlagsAndMode`?

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+  roi_relative_ = spec.GetArgument<bool>("roi_relative");
+
+  if (has_mode_) {
+    mode_ = ParseResizeMode(spec.GetArgument<std::string>("mode"));
+  } else {
+    mode_ = ResizeMode::Default;
+  }
+
+  if (has_axes_)
+    axes_ = spec.GetRepeatedArgument<int>("axes");
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:100]: Do you validate sanity of those values, like non-repeated values, in 0..ndim -1 range? Should they be monotone?

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+  int nargs = axes_.size();
+  int first_axis = ndim_;
+  int last_axis = -1;
+  for (int axis : axes_) {
+    if (axis < first_axis)
+      first_axis = axis;
+    if (axis > last_axis)
+      last_axis = axis;
+  }
+  spatial_ndim_ = last_axis - first_axis + 1;
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:146]: Can or should that be different from `axes_.size();`?

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+  max_size_.resize(spatial_ndim_,
+                   std::nextafter(static_cast<float>(std::numeric_limits<int>::max()), 0.0f));
+  if (has_max_size_) {
+    GetSingleOrRepeatedArg(spec, max_size_arg_, "max_size", nargs);
+    for (int i = 0; i < nargs; i++) {
+      int d = axes_[i] - first_spatial_dim_;
+      assert(d >= 0);
+      assert(d < spatial_ndim_);
+      max_size_[d] = max_size_arg_[i];
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:184]: Here, the `max_size_` has `spatial_ndim_` size while we may visit up to `nargs` different `d`s.

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+                                           const TensorListShape<> &input_shape) {
+  SetFlagsAndMode(spec);
+
+  ndim_ = input_shape.sample_dim();
+  if (ndim_ < 1)
+    throw std::invalid_argument("Expected at least 1D inputs");
+  if (!has_axes_) {
+    axes_.resize(ndim_);
+    std::iota(axes_.begin(), axes_.end(), 0);
+  }
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:136]: nitpick: I understand it has much to do with the fact that the params depend on one another but it is quite hard to follow what is already loaded, needs to be defaulted and if it is validated if user provided.

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 9 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:100]: I guess I can add a check for in-range and unique values. They don't need to be in ascending order though

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 10 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:146]: The thing is that the user can provide no-op arguments (keeping same size for some dims). In such case spatial_ndim_ and first_spatial_dim_ are tweaked later.

Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:146]: Also, there's the possibility to provide axes = (0, 2), in which case, spatian_ndim will be 3 (and not 2)

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 11 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:184]: yes, spatial_ndim_ is always >= nargs (we take the distance from the first provided axis to the last, while some of the axes in the middle may be omitted).

File: dali/operators/generic/resize/tensor_resize.h (hunk 7 above)
Comment (jantonguirao) [reply dali/operators/generic/resize/tensor_resize.h:60]: Done

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 12 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:158]: Do you have any suggestion about how to make it clearer?

File: dali/operators/generic/resize/tensor_resize.h [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+  using Operator<Backend>::RunImpl;
+
+  TensorResizeAttr resize_attr_;
+  ResamplingFilterAttr resampling_attr_;
+};
+
+template <typename Backend>
+TensorResize<Backend>::TensorResize(const OpSpec &spec)
+    : Operator<Backend>(spec), ResizeBase<Backend>(spec), resize_attr_(spec) {
+  resample_params_.resize(num_threads_);
```
Comment (stiepan) [inline dali/operators/generic/resize/tensor_resize.h:95]: Why do we need that resize here? You do next the:
```
resample_params_.resize(nsamples * NumSpatialDims());
```

Comment (jantonguirao) [reply dali/operators/generic/resize/tensor_resize.h:95]: Copied from the other resize, but I don't think it makes any sense here.

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    }
+  }
+
+  SmallVector<float, 3> requested_size, in_lo, in_hi;
+  requested_size.resize(spatial_ndim_);
+  in_lo.resize(spatial_ndim_);
+  in_hi.resize(spatial_ndim_);
+
+  assert(has_sizes_ + has_scales_ == 1);
+  const float *max_size = has_max_size_ ? max_size_.data() : nullptr;
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:220]: Maybe it is nitpicky, but could you move `max_size = nullptr` declaration to L204? There is the section that handles this param and then it is only used in `CalculateSampleParams`. Here, it is kind of interleaved with `requested_size, in_lo, in_hi`.

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        return false;
+      }
+    }
+    return true;
+  };
+
+  int trim_start_ndim = 0;
+  int trim_end_ndim = 0;
+  for (int d = 0; d < spatial_ndim_; d++) {
+    if (!unchanged_dim(d))
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:261]: Shouldn't that be `d + first_spatial_dim_`?

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+      break;
+    trim_start_ndim++;
+  }
+  for (int d = spatial_ndim_ - 1; d >= 0; d--) {
+    if (!unchanged_dim(d))
+      break;
+    trim_end_ndim++;
+  }
+
+  int new_spatial_ndim = spatial_ndim_ - (trim_start_ndim + trim_end_ndim);
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.cc:271]: Is it possible that we have a total no-op specified here and both `trim_start_ndim` and `trim_end_ndim` end up equal to spatial_ndim_?

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 15 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:261]: true

File: dali/operators/generic/resize/tensor_resize.h [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+  void RunImpl(Workspace &ws) override;
+
+  void PrepareParams(const ArgumentWorkspace &ws, const TensorListShape<> &input_shape,
+                     const TensorLayout &layout) {
+    int nsamples = input_shape.num_samples();
+    resize_attr_.PrepareResizeParams(spec_, ws, input_shape, layout);
+
+    if (NumSpatialDims() < 2) {
+      // PrepareResizeParams should expand to at least
```
Comment (stiepan) [inline dali/operators/generic/resize/tensor_resize.h:54]:
```suggestion
      // PrepareResizeParams should expand to at least [something]?
```

File: dali/operators/image/resize/tensor_resize_attr.h [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+  int FirstSpatialDim() const {
+    return first_spatial_dim_;
+  }
+
+  span<const ResizeParams> Params() const {
+    return make_cspan(params_);
+  }
+
+ private:
+  void PrepareAxes(const OpSpec &spec, int ndim);
```
Comment (stiepan) [inline dali/operators/image/resize/tensor_resize_attr.h:72]: Where is this defined and used? I can see call to `axes_helper_` method, but can't see this one.

Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.h]: Leftover. Removed

File: dali/operators/generic/resize/tensor_resize.h (hunk 17 above)
Comment (jantonguirao) [reply dali/operators/generic/resize/tensor_resize.h]: Done (the comment got moved too)

File: dali/operators/generic/resize/tensor_resize.h [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+      throw std::logic_error("1D resizing is not supported");
+    }
+
+    if (NumSpatialDims() < 1) {
+      throw std::invalid_argument(
+          "TensorResize expects at least one spatial dimension to be resized");
+    }
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/generic/resize/tensor_resize.h]: True, I'll simplify to a simple assert (should not happen)

**Review by stiepan — APPROVED** [review]

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+``out_size = size_rounding(scale_factor * original_size)``.
+See ``size_rounding`` for a list of supported rounding policies.
+
+When ``axes`` is provided, the scale factor values refer to the axes specified.
+Note: Arguments ``sizes`` and ``scales`` are mutually exclusive.)code",
+                                   {}, true)
+    .AddOptionalArg<float>("alignment", R"code(Determines the position of the ROI
+when using scales (provided or calculated).
+
+The real output size must be integer and may differ from "ideal" output size calculated as input
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:45]:
```suggestion
The real output size must be integral and may differ from "ideal" output size calculated as input
```
(or "an integer")

File: dali/operators/image/resize/tensor_resize_attr.h [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  }
+
+  TensorListShape<> ExpandedInputShape() const {
+    return input_shape_;
+  }
+
+  int LeadingDummyDims() const {
+    return add_leading_spatial_ndim_;
+  }
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.h:57]: I think this shouldn't be in TensorResizeAttr - this is a generic base class whereas the need for leading dummy dims is an implementation detail.

File: dali/operators/image/resize/tensor_resize_attr.cc [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+  span<int> axes = axes_helper_.Axes();
+  first_spatial_dim_ = 0;
+  // We need spatial_ndim >= 2
+  add_leading_spatial_ndim_ = std::max(0, 2 - ndim_);
+  if (add_leading_spatial_ndim_ > 0) {
+    expand_dims(input_shape_, input_shape, first_spatial_dim_, ndim_ + add_leading_spatial_ndim_);
+    ndim_ = input_shape_.sample_dim();
+    for (int &a : axes)  // skip dummy dims
+      a += add_leading_spatial_ndim_;
+  }
```
Comment (mzient) [inline dali/operators/image/resize/tensor_resize_attr.cc:200]: I think this logic should go to the `TensorResize` class - it's a limitation of a particular implementation, not a generic thing.

File: dali/operators/image/resize/tensor_resize_attr.h (hunk 21 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.h:57]: Done

File: dali/operators/image/resize/tensor_resize_attr.cc (hunk 22 above)
Comment (jantonguirao) [reply dali/operators/image/resize/tensor_resize_attr.cc:200]: Done

**Comment by jantonguirao** [comment]
!build

File: dali/operators/generic/resize/tensor_resize.h [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+    resize_attr_.PrepareResizeParams(spec_, ws, input_shape, layout);
+
+    // 1st: Expand number of dimensions to at least 2D
+    int min_spatial_ndim = 2;  // ResizeBase requires it.
+    int orig_spatial_ndim = resize_attr_.NumSpatialDims();
+    add_leading_spatial_ndim_ = std::max(0, min_spatial_ndim - orig_spatial_ndim);
+    first_spatial_dim_ = 0;
+    spatial_ndim_ = orig_spatial_ndim + add_leading_spatial_ndim_;
+    expand_dims(expanded_input_shape_, input_shape, first_spatial_dim_, spatial_ndim_);
+    (void) input_shape;  // should use expanded_input_shape_ from now on
```
Comment (mzient) [inline dali/operators/generic/resize/tensor_resize.h:111]: What is this line for? Does the compiler issue a warning when a variable is used after a void cast or sth? If not, leave just the comment.

Comment (jantonguirao) [reply dali/operators/generic/resize/tensor_resize.h:111]: it was more for "documentation", but I can remove it.

File: dali/operators/generic/resize/tensor_resize.h [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+        if (static_cast<int64_t>(resize_params_[i].dst_size[d]) != extent ||
+            static_cast<int64_t>(resize_params_[i].src_lo[d]) != 0 ||
+            static_cast<int64_t>(resize_params_[i].src_hi[d]) != extent) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    int can_trim_n = std::max(0, spatial_ndim_ - min_ndim);  // at least 2 spatial dims should remain
```
Comment (mzient) [inline dali/operators/generic/resize/tensor_resize.h:63]:
```suggestion
    int can_trim_n = std::max(0, spatial_ndim_ - min_ndim);  // at least min_ndim spatial dims should remain
```

File: dali/operators/generic/resize/tensor_resize.h [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+            static_cast<int64_t>(resize_params_[i].src_lo[d]) != 0 ||
+            static_cast<int64_t>(resize_params_[i].src_hi[d]) != extent) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    // at least min_ndim should remain
+    int can_trim_n = std::max(0, spatial_ndim_ - min_ndim);
```
Comment (mzient) [inline dali/operators/generic/resize/tensor_resize.h:64]: Nitpick:
`can_trim_n` reads like a boolean (can I trim `n` dimensions?)
```suggestion
    int ndim_to_trim = std::max(0, spatial_ndim_ - min_ndim);
```

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build
